### PR TITLE
Add docs for HostEmulator

### PR DESCRIPTION
**Pull request by anuraaga**

### CODE REVIEW

**Comment by mathetake** [comment]
so clearly after #315, the e2e on http_body became flaky... could you take a look? @anuraaga

**Comment by mathetake** [comment]
maybe simply increase the timeoue on require.Eventually would be fine

File: proxywasm/proxytest/proxytest.go [hunk 1]
```diff
@@ -24,49 +24,113 @@ import (
 	"github.com/tetratelabs/proxy-wasm-go-sdk/proxywasm/types"
 )
 
+// HostEmulator implements the host side of proxy-wasm. Methods on HostEmulator will either invoke
+// methods in the plugin or return state about the host itself, reflecting any changes from previous
+// plugin invocations.
 type HostEmulator interface {
 	// Root
```
Comment (mathetake) [inline proxywasm/proxytest/proxytest.go:31]:
```suggestion
```

File: proxywasm/proxytest/proxytest.go [hunk 2]
```diff
@@ -24,49 +24,113 @@ import (
...
+	// GetTickPeriod returns the current tick period in the host.
 	GetTickPeriod() uint32
+	// Tick executes types.PluginContext.OnTick in the plugin.
 	Tick()
+	// GetQueueSize gets the current size of the queue in the host.
 	GetQueueSize(queueID uint32) int
+	// RegisterForeignFunction registers the foreign function in the host.
 	RegisterForeignFunction(name string, f func([]byte) []byte)
 
 	// network
```
Comment (mathetake) [inline proxywasm/proxytest/proxytest.go:71]:
```suggestion
```

File: proxywasm/proxytest/proxytest.go [hunk 3]
```diff
@@ -24,49 +24,113 @@ import (
...
+	// CallOnDownstreamData executes types.TcpContext.OnDownstreamData in the plugin.
 	CallOnDownstreamData(contextID uint32, data []byte) types.Action
+	// CloseUpstreamConnection executes types.TcpContext.OnUpstreamClose in the plugin.
 	CloseUpstreamConnection(contextID uint32)
+	// CloseDownstreamConnection executes types.TcpContext.OnDownstreamClose in the plugin.
 	CloseDownstreamConnection(contextID uint32)
+	// CompleteConnection executes types.TcpContext.OnStreamDone in the plugin.
 	CompleteConnection(contextID uint32)
 
 	// http
```
Comment (mathetake) [inline proxywasm/proxytest/proxytest.go:86]:
```suggestion
```

**Review by mathetake — APPROVED** [review]
code lgtm

**Comment by anuraaga** [comment]
Thanks @mathetake - on my fork I could reproduce the failure after one retry, and so far after one retry with raised timeout it is passing. Let's see how it goes

**Comment by anuraaga** [comment]
Actually - curious, should we only have the `checkMessage` in eventually and move the rest out? I think the other parts are supposed to be deterministic

**Comment by mathetake** [comment]
LOL failed...

**Comment by anuraaga** [comment]
No go on the test still btw, will keep on trying

**Comment by mathetake** [comment]
https://github.com/tetratelabs/proxy-wasm-go-sdk/compare/testaaaa?expand=1 will give us more context

**Comment by mathetake** [comment]
ok somewhat deterministially `Test_http_body/request/append` failing

**Comment by mathetake** [comment]
I couldn't reproduce on my local ubuntu...

**Comment by mathetake** [comment]
I hope this fixes https://github.com/tetratelabs/proxy-wasm-go-sdk/pull/320 🤞

**Pull request merged**